CREATE NEW PROJECT IMPLEMENTATION

TarikVelic opened the pull request:
--------------------------------------------------------------------
 MAIN
 * Created a model for creating a new project
 * Added post route to add new project
 * Added get route to get the project 
 --------------------------------------------------------------------
 OTHER
 *
--------------------------------------------------------------------`

TarikVelic committed:
CREATE NEW PROJECT IMPLEMENTATION

--------------------------------------------------------------------
 MAIN
 * Created a model for creating a new project
 * Added post route to add new project
 * Added get route to get the project
--------------------------------------------------------------------
 OTHER
 *
--------------------------------------------------------------------`

becarselman commented on app.js:
```diff
@@ -30,4 +35,4 @@ app.get('/', (req, res) => {
 
 app.listen(port, async () => {
   console.log(`Express application is running on http://localhost:${port}`)
```
I know this is old code but we need to remove url, or add in env and than display. you can call it BaseUrl

becarselman commented on controllers/projects.js:
```diff
@@ -0,0 +1,24 @@
+  async createProject(req, res, next) {
+    try {
+      const { name, description, duration, developers, projectType, hourlyRate, projectValue, endDate, salesChannel, isFinished, } = req.body;
+      const project = await projectService.createProject(name, description, duration, developers, projectType, 
```
Author of the famous book "Clean code", Robert Cecil Martin said that "The best functions are those with no parameters".

becarselman replied:
instead of sending req.body. You send 10 parameters. You can send req.body and in service use it like data.name, data.description ... Also you need validation for this. @ElvaRivle create good example of validation in middleware so you can use his code as example.

becarselman commented on services/projects.js:
```diff
@@ -0,0 +1,32 @@
+
+
+module.exports = {
+  async createProject(name, description, duration, developers, projectType, hourlyRate, projectValue, endDate, salesChannel, isFinished,) {
```
this part is connected to comment with parameters

becarselman commented on services/projects.js:
```diff
@@ -0,0 +1,32 @@
+
+module.exports = {
+  async createProject(name, description, duration, developers, projectType, hourlyRate, projectValue, endDate, salesChannel, isFinished,) {
+    const project = new Project({
```
Please pay more attention I already comment this on previous pr. Everything that need to be done with database (create, delete, get, update) will go in repository. And than you can reuse it. You create good example of reusability in previous pr where you reuse Elvir's getUserByEmail function from repository

becarselman commented on services/projects.js:
```diff
@@ -0,0 +1,32 @@
+  
+    });
+    await project.validate();
+    await project.save();
```
use ORM

becarselman commented on services/projects.js:
```diff
@@ -0,0 +1,32 @@
+  },
+
+  async getProject(id) {
+    const project = await Project.findById(id);
```
same

becarselman commented on services/projects.js:
```diff
@@ -0,0 +1,32 @@
+  async getProject(id) {
+    const project = await Project.findById(id);
+    if (!project) {
+      throw new Error(errors.PROJECT_NOT_FOUND);
```
you can reuse general error message "not found"

becarselman commented on app.js:
```diff
@@ -5,21 +5,26 @@ const loginRoutes = require("./routes/login")
+}));
 app.use("/login", loginRoutes)
 app.use("/register", registerRoutes)
+app.use("/forgot-password", forgotPasswordRoutes)
```
?

becarselman commented on app.js:
```diff
@@ -5,21 +5,26 @@ const loginRoutes = require("./routes/login")
 app.use("/login", loginRoutes)
 app.use("/register", registerRoutes)
+app.use("/forgot-password", forgotPasswordRoutes)
+app.use("/reset-password", resetPasswordRoutes)
```
?

becarselman commented on app.js:
```diff
@@ -5,21 +5,26 @@ const loginRoutes = require("./routes/login")
 const sendMailRouter = require('./routes/sendemail')
 const registerRoutes = require("./routes/register")
-const {FRONTEND_URL} = require("./configuration/env");
+const forgotPasswordRoutes = require("./routes/forgotPassword")
```
?

becarselman commented on app.js:
```diff
@@ -5,21 +5,26 @@ const loginRoutes = require("./routes/login")
 const registerRoutes = require("./routes/register")
-const {FRONTEND_URL} = require("./configuration/env");
+const forgotPasswordRoutes = require("./routes/forgotPassword")
+const resetPasswordRoutes = require("./routes/resetPassword") 
```
?

TarikVelic committed:
FEATURE/TariikMVP/Project-v3

--------------------------------------------------------------------
 MAIN
 * Fixing stuff from review
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on repositories/projects.js:
```diff
@@ -0,0 +1,17 @@
+const Project = require('../db/models/projects');
+
+async function createProject(projectData) {
+  const project = new Project(projectData);
```
why you don't use Create function

TarikVelic committed:
small changes in code

becarselman commented on services/projects.js:
```diff
@@ -0,0 +1,19 @@
+const joi = require("joi");
```
unused import

becarselman commented on utils/validation_schemas/projects-data.js:
```diff
@@ -0,0 +1,65 @@
+        "string.empty": errors.PROJECT_NAME_NOT_PROVIDED,
+        "string.required": errors.PROJECT_NAME_NOT_PROVIDED,
+    }),
+    description: joi.string().required().messages({
```
I think joi will automaticly return error "description missing" if you just put required(). I think you don't need to specify error messages. But check it

TarikVelic committed:
FEATURE/TariikMVP/Project-v3

--------------------------------------------------------------------
 MAIN
 * Small changes in code, fixing stuff from reviews
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/Project-v3

--------------------------------------------------------------------
 MAIN
 * Importing joi error messages
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/Project-v3

--------------------------------------------------------------------
 MAIN
 * Changing error response status from 401 to 400
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
Small changes in code
